pkg/clustertree/cluster-manager/cluster_controller.go

```diff
@@ -0,0 +1,210 @@
+	c.ManagerCancelFuncs = make(map[string]*context.CancelFunc)
+	c.ControllerManagers = make(map[string]*manager.Manager)
+	c.Logger = mgr.GetLogger()
+
+	return controllerruntime.NewControllerManagedBy(mgr).
+		Named(ControllerName).
+		WithOptions(controller.Options{}).
+		For(&clusterlinkv1alpha1.Cluster{}, builder.WithPredicates(predicatesFunc)).
+		Complete(c)
+}
+
+func (c *ClusterController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	klog.V(4).Infof("============ %s starts to reconcile %s ============", ControllerName, request.Name)
+	defer func() {
+		klog.V(4).Infof("============ %s has been reconciled =============", request.Name)
+	}()
+
+	cluster := &clusterlinkv1alpha1.Cluster{}
+	if err := c.Master.Get(ctx, request.NamespacedName, cluster); err != nil {
+		if errors.IsNotFound(err) {
+			klog.Infof("Cluster %s has been deleted", request.Name)
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{}, err
+	}
+
+	// ensure finalizer
+	if cluster.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(cluster, ControllerFinalizerName) {
+			controllerutil.AddFinalizer(cluster, ControllerFinalizerName)
+			if err := c.Master.Update(ctx, cluster); err != nil {
+				return controllerruntime.Result{}, err
+			}
+		}
+	}
+
+	if !cluster.DeletionTimestamp.IsZero() {
+		c.clearClusterControllers(cluster)
+
+		if controllerutil.ContainsFinalizer(cluster, ControllerFinalizerName) {
+			controllerutil.RemoveFinalizer(cluster, ControllerFinalizerName)
+			if err := c.Master.Update(ctx, cluster); err != nil {
+				return controllerruntime.Result{}, err
+			}
+		}
+	}
+
+	// cluster added or kubeconfig changed
+	c.clearClusterControllers(cluster)
+
+	// build mgr for cluster
+	config, err := utils.NewConfigFromBytes(cluster.Spec.Kubeconfig, func(config *rest.Config) {
+		config.QPS = DefaultClusterKubeQPS
+		config.Burst = DefalutClusterKubeBurst
+	})
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not build clientset for cluster %s: %v", cluster.Name, err)
+	}
+
+	mgr, err := controllerruntime.NewManager(config, controllerruntime.Options{
+		Logger:                 c.Logger.WithName("cluster-controller-manager"),
+		Scheme:                 scheme.NewSchema(),
+		LeaderElection:         false,
+		MetricsBindAddress:     "0",
+		HealthProbeBindAddress: "0",
+	})
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("new manager with err %v, cluster %s", err, cluster.Name)
+	}
+
+	subContext, cancel := context.WithCancel(ctx)
+
+	c.ControllerManagersLock.Lock()
+	c.ControllerManagers[cluster.Name] = &mgr
+	c.ManagerCancelFuncs[cluster.Name] = &cancel
+	c.ControllerManagersLock.Unlock()
+
+	if err = c.setupControllers(&mgr); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to setup cluster %s controllers: %v", cluster.Name, err)
+	}
+
+	go func() {
+		if err := mgr.Start(subContext); err != nil {
```

Comment:
controller on controller ？ what does it mean ？

Reply:
Yes, when the cluster joins, start the controller manager of the cluster

Reply:
> Yes, when the cluster joins, start the controller manager of the cluster

One member cluster will have a single controller manager?

Reply:
yes